### PR TITLE
[WIP] Progress check on Piezo Changes

### PR DESCRIPTION
**Pull request by mcdonnelldean**
@lyzadanger @Resseguie @julianduque @BrianGenisio 

Hi all,

I have tidied up this PR some what clarifying what is proposed. Currently the new public API is
fully implemented but needs more tests. There is a lump of logic that we may be able to remove
from the internal API if we include Octavian to handle all logic except the handling of melodies.
## Public API

The new public API assumes that we are using Octavian (talked about below). By using Octavian the 
new api becomes drastically simplified. The same constructor is retained and allows the provision
of default values.
- Piezo(opts)
- .pulsepin(pulse, duration)
- .playFrequency(frequency, duration)
- .playNote(note, duration)
- .playMelody(melody, callback)
- .stop()
## RTTF

Ring Tone Transfer Format or RTTF is a language that was designed for typing melodies into 
cellphones via a keypad. It's simple format and pre-existing database of sounds make this the
perfect input format to enable more advanced melodies to be played via the Piezo class.
### Sample RTTF Input

```
'd=8 o=4 b=125: f, a, 4b, f, a, 4b, f, a, 4b, e5, 4d5, b, c5, b, g, 2e, a, b, g, 2e, p2'
```

The input string has a header section, broken down as follows:
- _d=_ - The default beat for notes who's beat is not specified.
- _o=_ - The default octave for notes who's octave is not specified.
- _b=_ - The overall tempo of the melody.

After the colon each note in the melody is specified with at least the note value and optionally
the octave and beat, where this is not provided the defaults are used.

RTTF Has some more advanced features that need to be built in, including repeating and half notes,
these are not currently supported but will be before the PR goes live.
### RTTF+

```
'b=8 o=4 t=125: f a 4b f a 4b f a 4b e5 4d5 b c5 b g 2e a b g 2e p2
```

RTTF+ is a simplification of RTTF, it removes the need to use comma's as well as fixes the header
naming scheme.
## Octavian

[See #809] [Octavian](https://github.com/stevekinney) is a lib written by  [Steve Kinney](https://github.com/stevekinney). It manages Notes and all of the infrastructure around 
them. Usage of Octavian would most likely rest on having the following features implemented.
- RTTF input support for notes
- Beats scale, beats on notes, beat normalization
- (optionally) Pulse calculation

Because of the logic around beats and parsing we would need the first two supported as a minimum
otherwise we will need to add more parsing logic to handle RTTF to Octavian format.

@SteveKinney It would be awesome to discuss if any of these changes would make sense for your
library. I can help with the changes but it's conditional on a roughly 3 week lead time as I
am swamped. I'm available for skype or I.M/mail though to help whenever.

### CODE REVIEW

File: lib/piezo.js [hunk 1]
```diff
@@ -1,150 +1,398 @@
...
+  "f8":  5588,
   "f#8": 5920,
-  "g8": 6272,
+  "g8":  6272,
   "g#8": 6645,
-  "a8": 7040,
+  "a8":  7040,
   "a#8": 7459,
-  "b8": 7902
+  "b8":  7902
```
Comment (rwaldron) [inline lib/piezo.js]: Just run the file through `grunt jsbeautifier`

**Comment by rwaldron** [comment]
A lot of time and discussion between several contributors went into the existing API. I urge you to work with them on this.

File: lib/piezo.js [hunk 2]
```diff
@@ -1,150 +1,398 @@
...
-function clearTimer() {
-  if (!this.timer) {
-    return;
+Piezo.Beats = {
+  "second":   ["2", 2],
+  "fourth":   ["4", 1],
+  "eight":    ["8", 1/2],
+  "sixteenth":["16", 1/4]
+};
+
```
Comment (rwaldron) [inline lib/piezo.js]: Why is the first value a string of a number?

File: lib/piezo.js [hunk 3]
```diff
@@ -1,150 +1,398 @@
...
+  // Ensure we have defaults properly set up.
+  melody.notes = melody.notes = [];
+  melody.octave = melody.octave || defaults.octave;
+  melody.beat = melody.beat || defaults.beats;
+  melody.tempo = melody.tempo || defaults.tempo;
+
+  // The sheet can contain defaults for beat (b=), octave (o=),
+  // and tempo (t=). They always come before the first note.
+  var sheetElements  = melody.sheet.match(/\S+/g) || [];
+  _.forEach(sheetElements, function(element) {
```
Comment (rwaldron) [inline lib/piezo.js]: Is sheetElements an Array? If yes, then use sheetElements.forEach

File: lib/piezo.js [hunk 4]
```diff
@@ -1,150 +1,398 @@
...
+    var octavedNote = note.notation + note.octave;
+    if (Piezo.Notes[octavedNote]) {
+      frequency =  Piezo.Notes[octavedNote];
+    }
+  }
+
+  return frequency;
+};
+
+Piezo.prototype.pulsePin = function(pulse, duration) {
```
Comment (rwaldron) [inline lib/piezo.js]: Can this be just `pulse`?

File: lib/piezo.js [hunk 5]
```diff
@@ -1,150 +1,398 @@
...
+  this.isPlaying = true;
+  playNextNote();
+};
+
+Piezo.prototype.clearTimer = function() {
+  this.io.digitalWrite(this.pin, 0);
+  clearTimer.call(this);
+
+  return this;
+};
```
Comment (rwaldron) [inline lib/piezo.js]: Why is this exposed?

File: lib/piezo.js [hunk 6]
```diff
@@ -1,150 +1,398 @@
...
+        normalizedNote.duration = note[0];
+        normalizedNote.notation = note[1];
+      }
+
+      return normalizedNote;
+    }
+  }
+
+  return null;
+};
```
Comment (rwaldron) [inline lib/piezo.js]: This is basically a constructor. Why not just define a private Note class?

**Comment by McDonnellDean** [comment]
@rwaldron absolutely! That's why ai have it here in the rough and why I kept the old api in tact. I'm open to any suggestions even keeping the existing bits and merging functionality. I'd love to get some feedback and this early before getting into finalising anything.

**Comment by rwaldron** [comment]
I'll take a closer look in the AM, but everyone else should review as well.

**Comment by McDonnellDean** [comment]
Oh no! I'm missing some sample code and a description of pulsePin() which is really just note() named more like its function. I'll fix this up first thing in the morning.

File: lib/piezo.js (hunk 2 above)
Comment (McDonnellDean) [inline lib/piezo.js]: I incorrectly assumed I couldn't have the number part as the first part. Will change

File: lib/piezo.js (hunk 3 above)
Comment (McDonnellDean) [inline lib/piezo.js]: Done

File: lib/piezo.js (hunk 4 above)
Comment (McDonnellDean) [inline lib/piezo.js]: yes, it makes more sense like this because it's opposite is just called stop

File: lib/piezo.js (hunk 5 above)
Comment (McDonnellDean) [inline lib/piezo.js]: Wrong name, should be stop

File: lib/piezo.js (hunk 6 above)
Comment (McDonnellDean) [inline lib/piezo.js]: Is there a need for it? This is the only place it would be used

File: lib/piezo.js [hunk 7]
```diff
@@ -129,22 +137,274 @@ Piezo.Notes = {
...
+    if (beatElement === beat.toString()) {
+      normalizedBeat = Piezo.Beats[beatElement];
+    }
+  });
+
+  return normalizedBeat;
+};
+
+Piezo.normalizeOctave = function(octave) {
+  if (octave === "" || isNaN(octave) || octave < 0 || octave > 8) {
```
Comment (rwaldron) [inline lib/piezo.js]: `isNaN` => `Number.isNaN`

File: lib/piezo.js [hunk 8]
```diff
@@ -129,202 +137,284 @@ Piezo.Notes = {
...
+    return null;
   }
-  if (typeof tone === "string") {
-    tone = tone.toLowerCase();
 
-    var needsOctave = isNaN(tone.substr(tone.length -1));
-    if (needsOctave) {
-      tone = tone + Piezo.prototype.defaultOctave().toString();
+  var normalizedBeat = null;
+  _.forEach(Object.keys(Piezo.Beats), function(beatElement) {
```
Comment (rwaldron) [inline lib/piezo.js]: `_.forEach(Object.keys(Piezo.Beats)` => `Object.keys(Piezo.Beats).forEach(`

File: lib/piezo.js [hunk 9]
```diff
@@ -129,202 +137,284 @@ Piezo.Notes = {
...
+  // Ensure we have defaults properly set up.
+  melody.notes = melody.notes = [];
+  melody.octave = melody.octave || defaults.octave;
+  melody.beat = melody.beat || defaults.beats;
+  melody.tempo = melody.tempo || defaults.tempo;
+
+  // The sheet can contain defaults for beat (b=), octave (o=),
+  // and tempo (t=). They always come before the first note.
+  var sheetElements = melody.sheet.match(/\S+/g) || [];
+  _.forEach(sheetElements, function(element) {
```
Comment (rwaldron) [inline lib/piezo.js]: ` _.forEach(sheetElements` => `sheetElements.forEach(`

File: lib/piezo.js [hunk 10]
```diff
@@ -129,202 +137,284 @@ Piezo.Notes = {
...
+    if ((element.indexOf("b=") > -1 && melody.sheet.indexOf("t=") > -1) ||
```
Comment (rwaldron) [inline lib/piezo.js]: Use `element.includes()`

File: lib/piezo.js [hunk 11]
```diff
@@ -129,202 +137,284 @@ Piezo.Notes = {
...
-  return this;
-};
+    if (element.indexOf("o=") > -1) {
+      var octaveValue = element.substring(2);
+      return melody.octave = normalizeOctave(octaveValue) || melody.octave;
+    }
+
+    if ((element.indexOf("t=") > -1 && melody.sheet.indexOf("b=") > -1) ||
+        (element.indexOf("b=") > -1 && melody.sheet.indexOf("d=") > -1)) {
+      var tempoValue = element.substring(2);
```
Comment (rwaldron) [inline lib/piezo.js]: Use `element.includes()`

File: lib/piezo.js [hunk 12]
```diff
@@ -129,202 +137,284 @@ Piezo.Notes = {
...
-    lastNote = song[song.length - 1];
-    if (lastNote && lastNote[0] === note) {
-      lastNote[1] += beats;
-    } else {
-      song.push([note, beats]);
+function noteToPulse(note) {
+  var frequency = noteToFrequency(note);
+  var pulse = frequencyToPulse(frequency);
+
+  return pulse;
```
Comment (rwaldron) [inline lib/piezo.js]: `return frequencyToPulse(noteToFrequency(note));`

File: lib/piezo.js [hunk 13]
```diff
@@ -129,202 +137,284 @@ Piezo.Notes = {
...
-  next();
+  playNextNote();
 
   return this;
 };
 
-Piezo.prototype.noTone = function() {
+Piezo.prototype.stop = function() {
   this.io.digitalWrite(this.pin, 0);
   clearTimer.call(this);
```
Comment (rwaldron) [inline lib/piezo.js]: Let's change this to match `Led`: 

- `stop` just stops an internal timer, but doesn't change the state 
- `off` changes the state to off

It's a slight change from before, but we should be ok with breaking small things in the name of consistency.

File: lib/piezo.js [hunk 14]
```diff
@@ -129,202 +137,284 @@ Piezo.Notes = {
...
+
+  // 1. remove colons and semi-colons
+  // 2. remove additional whitespace
+  // 3. clean up defaults if present.
+  melody.sheet = melody.sheet.replace(/,|:/g, " ");
+  melody.sheet = melody.sheet.replace(/\s{2,}/g, " ");
+  melody.sheet = melody.sheet.replace(/\s{1,}=|=\s{1,}/g, "=");
+
+  // Ensure we have defaults properly set up.
+  melody.notes = melody.notes = [];
```
Comment (rwaldron) [inline lib/piezo.js]: `melody.notes = melody.notes = [];`? Should this be: `melody.notes = melody.notes || [];`?

File: test/piezo.js [hunk 15]
```diff
@@ -181,222 +192,55 @@ exports["Piezo"] = {
...
-        tempo: tempo
-      };
-    test.expect(1);
-
-    this.piezo.play(tune, function() {
-      test.ok(1); // We made it this far, no choking on bad values
-      test.done();
-    }.bind(this));
-
-  },
```
Comment (rwaldron) [inline test/piezo.js]: I'm concerned about all of these deleted tests.

Comment (mcdonnelldean) [inline test/piezo.js]: There are a lot more tests to come, i just have the bear minimum to get the octvian work in.

File: lib/piezo.js (hunk 13 above)
Comment (mcdonnelldean) [inline lib/piezo.js]: Noted

File: lib/piezo.js (hunk 8 above)
Comment (mcdonnelldean) [inline lib/piezo.js]: Done as of next commit

File: lib/piezo.js (hunk 14 above)
Comment (mcdonnelldean) [inline lib/piezo.js]: Fixed on next commit

**Comment by mcdonnelldean** [comment]
@All,

I'll get this PR cleaned up and finalised for review in the next 10 days.

**Comment by rwaldron** [comment]
@Resseguie ^^

**Comment by mcdonnelldean** [comment]
Ugh, I'm super sorry! I completely forgot this. So, is it still valid (apart from cleanup). I need to re familiarise myself with the code again but is the way this was done still the 'correct' way to do this? I've seen some talk around packs or plugins for J5.
